core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -329,7 +330,17 @@ protected void validateJoinFilters(List<RexNode> aboveFilters,
     final Iterator<RexNode> filterIter = joinFilters.iterator();
     while (filterIter.hasNext()) {
       RexNode exp = filterIter.next();
-      if (!predicate.apply(join, joinType, exp)) {
+      int maxIndex = -1;
+      ImmutableList<RexInputRef> inputRefs = RexUtil.getAllInputRef(exp);
+      for (RexInputRef rexInputRef: inputRefs) {
+        if (rexInputRef.getIndex() > maxIndex) {
+          maxIndex = rexInputRef.getIndex();
+        }
+      }
```

Comment:
How about this:
```
maxIndex = RelOptUtil.InputFinder.bits(exp).length() - 1;
```
And the computation of maxIndex should be moved inside `if (!predicate.apply(join, joinType, exp)) {` to avoid unnecessary computation.

Reply:
Agree, that's better.

Reply:
Done